contracts/TokenVesting.sol

```diff
@@ -2,355 +2,442 @@
+        }
+        unchecked {
+            uint112 remainingAmount = vestingSchedule.amountTotal -
+                vestingSchedule.released;
+            vestingSchedulesTotalAmount -= remainingAmount;
+            vestingSchedule.revoked = true;
+            emit Revoked(vestingScheduleId, remainingAmount);
+        }
     }
 
     /**
-    * @dev Computes the vesting schedule identifier for an address and an index.
-    */
-    function computeVestingScheduleIdForAddressAndIndex(address holder, uint256 index)
-        public
-        pure
-        returns(bytes32){
-        return keccak256(abi.encodePacked(holder, index));
+     * @notice Withdraw the specified amount if possible.
+     * @param amount the amount to withdraw
+     */
+    function withdraw(uint256 amount) external onlyOwner {
+        if (getWithdrawableAmount() < amount) revert NotWithdrawable();
+        token.safeTransfer(owner(), amount);
+        emit Withdrawn(owner(), amount);
     }
 
     /**
-    * @dev Computes the releasable amount of tokens for a vesting schedule.
-    * @return the amount of releasable tokens
-    */
-    function _computeReleasableAmount(VestingSchedule memory vestingSchedule)
-    internal
-    view
-    returns(uint256){
-        uint256 currentTime = getCurrentTime();
-        if ((currentTime < vestingSchedule.cliff) || vestingSchedule.revoked == true) {
-            return 0;
-        } else if (currentTime >= vestingSchedule.start.add(vestingSchedule.duration)) {
-            return vestingSchedule.amountTotal.sub(vestingSchedule.released);
-        } else {
-            uint256 timeFromStart = currentTime.sub(vestingSchedule.start);
-            uint secondsPerSlice = vestingSchedule.slicePeriodSeconds;
-            uint256 vestedSlicePeriods = timeFromStart.div(secondsPerSlice);
-            uint256 vestedSeconds = vestedSlicePeriods.mul(secondsPerSlice);
-            uint256 vestedAmount = vestingSchedule.amountTotal.mul(vestedSeconds).div(vestingSchedule.duration);
-            vestedAmount = vestedAmount.sub(vestingSchedule.released);
-            return vestedAmount;
+     * @notice Release vested amount of tokens.
+     * @param vestingScheduleId the vesting schedule identifier
+     * @param amount the amount to release
+     */
+    function release(
+        bytes32 vestingScheduleId,
+        uint112 amount
+    ) public onlyIfVestingScheduleNotRevoked(vestingScheduleId) {
+        VestingSchedule storage vestingSchedule = vestingSchedules[
+            vestingScheduleId
+        ];
+        address sender = msg.sender;
+        address beneficiary = vestingSchedule.beneficiary;
+        bool isBeneficiary = sender == beneficiary;
+        bool isOwner = sender == owner();
+        if (!isBeneficiary && !isOwner) revert OnlyBeneficiaryOrOwner();
+        uint112 releasableAmount = _computeReleasableAmount(vestingSchedule);
+        amount = releasableAmount < amount ? releasableAmount : amount;
```

Comment:
Return error if it's requesting to release more than available?

Reply:
Why returning error instead of releasing what's available?